custom/abt/reports/expressions.py

```diff
@@ -1,15 +1,213 @@
+import os
 from jsonobject import JsonObject
+import yaml
 from corehq.apps.userreports.specs import TypeProperty
+from corehq.apps.app_manager.models import Application
+from dimagi.utils.decorators.memoized import memoized
 
 
 class AbtSupervisorExpressionSpec(JsonObject):
-    # TODO:
-    # This class is a stub added to demonstrate UCR expression plugin usage
-    # This class will be filled out in a PR coming soon - NC 4/20/15
     type = TypeProperty('abt_supervisor')
+    _questions_cache = {}
+
+    @property
+    @memoized
+    def _flag_specs(self):
+        """
+        Return a dict where keys are form xmlns and values are lists of FlagSpecs
+        """
+        path = os.path.join(os.path.dirname(__file__), 'flagspecs.yaml')
+        with open(path) as f:
+            return yaml.load(f)
+
+    @classmethod
+    def _get_val(cls, item, path):
+        """
+        Return the answer in the given submitted form (item) to the question specified by path.
+        Return empty tuple if no answer was given to the given question.
+        """
+        if path:
+            try:
+                v = item['form']
+                for key in path:
+                    v = v[key]
+                return v
+            except KeyError:
+                return ()
+
+    @classmethod
+    def _question_answered(cls, value):
+        """
+        Return true if the given value indicates that an answer was provided for its question.
+        """
+        return value != ()
+
+    @classmethod
+    def _raise_for_any_answer(cls, danger_value):
+        """
+        Return true if the given danger_value indicates that any question answer should raise the flag.
+        """
+        return danger_value == []
+
+    @classmethod
+    def _get_question_options(cls, item, question_path):
+        """
+        Return a list of option values for the given question path and item
+        (which is a dict representation of an XFormInstance)
+        """
+        app_id, xmlns = item['app_id'], item['xmlns']
+        questions = cls._questions_cache.get((app_id, xmlns), None)
```

Comment:
given that this is a classmethod I think this cache will live for essentially the lifecycle of the process, right? on the one hand that's great from a caching/speed perspective, though it may have unexpected behavior:
1. memory-wise - are there lotso of app_id/xmlns pairs? if so this may grow to be a high footprint attached to every python process.
2. if you change the app, the cache won't change until the process restarts (e.g. gunicorn serves a certain number of requests, pillowtop restarts, there's a deploy, etc.). I think that's fine since app changes might require table rebuilds anyways, just noting.

generally am interested how you made this design decision as opposed to using something much simpler like `quickcache`